File: src/main/java/com/example/complaint/Complaint.java
package com.example.complaint;

import com.example.coreapi.ComplaintFiledEvent;
import com.example.coreapi.FileComplaintCommand;
import org.axonframework.commandhandling.CommandHandler;
import org.axonframework.commandhandling.model.AggregateIdentifier;
import org.axonframework.eventsourcing.EventSourcingHandler;
import org.axonframework.spring.stereotype.Aggregate;

import static org.axonframework.commandhandling.model.AggregateLifecycle.apply;

@Aggregate
public class Complaint {

    @AggregateIdentifier
    private String complaintId;

    public Complaint() {
    }

    @CommandHandler
    public Complaint(FileComplaintCommand cmd) {
        apply(new ComplaintFiledEvent(cmd.getComplaintId(), cmd.getDescription(), cmd.getCompany()));
    }

    @EventSourcingHandler
    protected void on(ComplaintFiledEvent event) {
        this.complaintId = event.getComplaintId();
    }

}

File: src/main/java/com/example/coreapi/complaint.kt
package com.example.coreapi

class FileComplaintCommand(val complaintId : String, val description : String, val company : String)
class ComplaintFiledEvent(val complaintId : String, val description : String, val company : String)

class ResolveComplaintCommand(val complaintId: String, val satisfactionIndex : Int)
class ComplaintResolvedEvent(val complaintId: String, val satisfactionIndex : Int)
